ranger-discovery-bundle/src/main/java/io/appform/ranger/discovery/bundle/id/nonce/NonceGeneratorBase.java

```diff
@@ -0,0 +1,136 @@
+package io.appform.ranger.discovery.bundle.id.nonce;
+
+import com.google.common.base.Preconditions;
+import io.appform.ranger.discovery.bundle.id.Domain;
+import io.appform.ranger.discovery.bundle.id.Id;
+import io.appform.ranger.discovery.bundle.id.IdInfo;
+import io.appform.ranger.discovery.bundle.id.IdValidationState;
+import io.appform.ranger.discovery.bundle.id.constraints.IdValidationConstraint;
+import io.appform.ranger.discovery.bundle.id.formatter.IdFormatter;
+import io.appform.ranger.discovery.bundle.id.formatter.IdFormatters;
+import io.appform.ranger.discovery.bundle.id.generator.IdGeneratorBase;
+import io.appform.ranger.discovery.bundle.id.request.IdGenerationRequest;
+import lombok.Getter;
+import lombok.val;
+import org.joda.time.DateTime;
+
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+
+@Getter
+public abstract class NonceGeneratorBase {
+
+    private final SecureRandom SECURE_RANDOM = new SecureRandom(Long.toBinaryString(System.currentTimeMillis()).getBytes());
+    private final List<IdValidationConstraint> GLOBAL_CONSTRAINTS = new ArrayList<>();
+    private final Map<String, Domain> REGISTERED_DOMAINS = new ConcurrentHashMap<>(Map.of(Domain.DEFAULT_DOMAIN_NAME, Domain.DEFAULT));
+    private final IdFormatter idFormatter;
+
+    protected NonceGeneratorBase(final IdFormatter idFormatter) {
+        this.idFormatter = idFormatter;
+    }
+
+    public synchronized void cleanUp() {
```

Comment:
make methods that you don't want to be overridden as final in base class. Only allow controlled override